### PR TITLE
Original caller

### PR DESCRIPTION
**Pull request by matei-p**

### CODE REVIEW

File: mock/context/outputContextStub.go [hunk 1]
```diff
@@ -159,9 +159,9 @@ func (o *OutputContextStub) Transfer(destination []byte, sender []byte, gasLimit
 }
 
 // TransferESDT mocked method
-func (o *OutputContextStub) TransferESDT(destination []byte, sender []byte, transfers []*vmcommon.ESDTTransfer, callInput *vmcommon.ContractCallInput) (uint64, error) {
+func (o *OutputContextStub) TransferESDT(destination []byte, originalCaller []byte, sender []byte, transfers []*vmcommon.ESDTTransfer, callInput *vmcommon.ContractCallInput) (uint64, error) {
```
Comment (sasurobert) [inline mock/context/outputContextStub.go:162]: let's define some structure for this input - like ArgsTransfer or something like that. Too many arguments in input - it is a general refactor - can be done in the next PR.

File: vmhost/contexts/runtime.go [hunk 2]
```diff
@@ -480,6 +480,7 @@ func (context *runtimeContext) SetVMInput(vmInput *vmcommon.ContractCallInput) {
 	if len(vmInput.CallerAddr) > 0 {
 		context.vmInput.CallerAddr = make([]byte, len(vmInput.CallerAddr))
 		copy(context.vmInput.CallerAddr, vmInput.CallerAddr)
+		context.vmInput.OriginalCallerAddr = context.vmInput.CallerAddr
```
Comment (sasurobert) [inline vmhost/contexts/runtime.go:483]: OriginalCaller should be here as well. and do like a deep copy. - outside of the if.

File: vmhost/interface.go [hunk 3]
```diff
@@ -41,7 +41,7 @@ type VMHost interface {
 	Storage() StorageContext
 	EnableEpochsHandler() vmcommon.EnableEpochsHandler
 
-	ExecuteESDTTransfer(destination []byte, sender []byte, esdtTransfers []*vmcommon.ESDTTransfer, callType vm.CallType) (*vmcommon.VMOutput, uint64, error)
+	ExecuteESDTTransfer(destination []byte, originalCaller []byte, sender []byte, esdtTransfers []*vmcommon.ESDTTransfer, callType vm.CallType) (*vmcommon.VMOutput, uint64, error)
```
Comment (sasurobert) [inline vmhost/interface.go:44]: ArgStructure as input in the next PR.

File: vmhost/interface.go [hunk 4]
```diff
@@ -221,7 +222,7 @@ type OutputContext interface {
 	WriteLog(address []byte, topics [][]byte, data []byte)
 	TransferValueOnly(destination []byte, sender []byte, value *big.Int, checkPayable bool) error
 	Transfer(destination []byte, sender []byte, gasLimit uint64, gasLocked uint64, value *big.Int, asyncData []byte, input []byte, callType vm.CallType) error
-	TransferESDT(destination []byte, sender []byte, transfers []*vmcommon.ESDTTransfer, callInput *vmcommon.ContractCallInput) (uint64, error)
+	TransferESDT(destination []byte, originalCaller []byte, sender []byte, transfers []*vmcommon.ESDTTransfer, callInput *vmcommon.ContractCallInput) (uint64, error)
```
Comment (sasurobert) [inline vmhost/interface.go:225]: ArgStructure as input in the next PR.

File: vmhost/vmhooks/baseOps.go [hunk 5]
```diff
@@ -2690,8 +2700,11 @@ func ExecuteReadOnlyWithTypedArguments(
 	metering.UseAndTraceGas(gasToUse)
 
 	sender := runtime.GetContextAddress()
+	originalCaller := runtime.GetOriginalCallerAddress()
+
 	contractCallInput, err := prepareIndirectContractCallInput(
 		host,
+		originalCaller,
```
Comment (sasurobert) [inline vmhost/vmhooks/baseOps.go:2707]: can't you have this inside the prepareIndirectContractCallInput ? from host you have access to runtime and GetOriginalCallerAddress

File: mock/context/outputContextStub.go (hunk 1 above)
Comment (matei-p) [reply mock/context/outputContextStub.go]: fixed

File: vmhost/vmhooks/managedei.go [hunk 6]
```diff
@@ -729,7 +730,7 @@ func (context *VMHooksImpl) ManagedCreateContract(
 	}
 
 	lenReturnData := len(host.Output().ReturnData())
-	newAddress, err := createContract(sender, data, value, metering, gas, code, codeMetadata, host, runtime)
+	newAddress, err := createContract(originalCaller, sender, data, value, metering, gas, code, codeMetadata, host, runtime)
```
[the remark this answers is not in the document]
Comment (matei-p) [reply vmhost/vmhooks/managedei.go]: fixed

File: vmhost/vmhooks/baseOps.go (hunk 5 above)
Comment (matei-p) [reply vmhost/vmhooks/baseOps.go]: fixed

File: vmhost/interface.go (hunk 4 above)
Comment (matei-p) [reply vmhost/interface.go]: fixed

File: vmhost/contexts/runtime.go (hunk 2 above)
Comment (matei-p) [reply vmhost/contexts/runtime.go]: fixed

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

File: go.mod [hunk 7]
```diff
@@ -52,3 +52,5 @@ require (
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+replace github.com/multiversx/mx-chain-vm-common-go v1.3.37-0.20230209123632-f0fbff86b763 => /home/bogdan/Elrond/elrond-vm-common
```
Comment (camilbancioiu) [inline go.mod:56]: Consider creating a branch on `mx-chain-vm-common-go` and using its commit hash instead.

File: vmhost/contexts/async.go [hunk 8]
```diff
@@ -150,14 +154,15 @@ func (context *asyncContext) InitStateFromInput(input *vmcommon.ContractCallInpu
...
 	newState := &asyncContext{
-		callID:          context.callID,
-		callerCallID:    context.callerCallID,
-		callerAddr:      context.callerAddr,
-		callback:        context.callback,
-		callbackData:    context.callbackData,
-		gasAccumulated:  context.gasAccumulated,
-		returnData:      context.returnData,
-		asyncCallGroups: context.asyncCallGroups, // TODO matei-p use cloneCallGroups()?
+		originalCallerAddr: context.originalCallerAddr,
```
Comment (camilbancioiu) [inline vmhost/contexts/async.go:157]: This needs to be added to `PopSetActiveState()` as well.

Comment (camilbancioiu) [reply vmhost/contexts/async.go:157]: And to `Clone()`.

File: vmhost/contexts/async.go [hunk 9]
```diff
@@ -108,6 +111,7 @@ func (context *asyncContext) InitState() {
 // information provided by a ContractCallInput.
 func (context *asyncContext) InitStateFromInput(input *vmcommon.ContractCallInput) error {
 	context.InitState()
+	context.originalCallerAddr = input.OriginalCallerAddr
```
Comment (camilbancioiu) [inline vmhost/contexts/async.go:114]: Copy instead of assigning the pointer.

**Review by camilbancioiu — CHANGES_REQUESTED** [review]

File: vmhost/contexts/async.go (hunk 9 above)
Comment (matei-p) [reply vmhost/contexts/async.go]: fixed

File: vmhost/contexts/async.go (hunk 8 above)
Comment (matei-p) [reply vmhost/contexts/async.go:160]: fixed

File: go.mod (hunk 7 above)
Comment (matei-p) [reply go.mod]: fixed

**Review by camilbancioiu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**